src/threads/ThreadState.cpp

```diff
@@ -1,22 +1,109 @@
+    {                                                                          \
+        if (m.find(id) == m.end()) {                                           \
+            faabric::util::UniqueLock lock(sharedMutex);                       \
+            if (m.find(id) == m.end()) {                                       \
+                m[id] = std::make_shared<T>(__VA_ARGS__);                      \
+            }                                                                  \
+        }                                                                      \
+    }                                                                          \
+    std::shared_ptr<T> varName = m[id];
+
 namespace threads {
 
-OpenMPContext& getOpenMPContext()
+static thread_local std::shared_ptr<Level> currentLevel = nullptr;
+
+std::mutex sharedMutex;
+
+std::unordered_map<uint32_t, std::shared_ptr<faabric::util::Barrier>> barriers;
+
+std::unordered_map<uint32_t, std::shared_ptr<std::recursive_mutex>>
+  levelMutexes;
+
+std::unordered_map<uint32_t, std::shared_ptr<std::mutex>> nowaitMutexes;
+std::unordered_map<uint32_t, std::shared_ptr<std::atomic<int>>> nowaitCounts;
+std::unordered_map<uint32_t, std::shared_ptr<std::condition_variable>>
+  nowaitCvs;
+
+void clearThreadState()
 {
-    static thread_local OpenMPContext ctx;
-    return ctx;
+    barriers.clear();
+
+    levelMutexes.clear();
+
+    nowaitMutexes.clear();
+    nowaitCounts.clear();
+    nowaitCvs.clear();
+}
+
+void setCurrentOpenMPLevel(const std::shared_ptr<Level>& level)
+{
+    currentLevel = level;
+}
+
+void setCurrentOpenMPLevel(
+  const std::shared_ptr<faabric::BatchExecuteRequest> req)
+{
+    if (req->contextdata().empty()) {
+        throw std::runtime_error("Empty context for OpenMP request");
+    }
+
+    currentLevel = levelFromBatchRequest(req);
+}
+
+std::shared_ptr<Level> getCurrentOpenMPLevel()
+{
+    if (currentLevel == nullptr) {
+        int nThreads =
+          faabric::scheduler::getScheduler().getThisHostResources().slots();
+        faabric::util::getLogger()->debug(
+          "Creating default OpenMP level with {} threads", nThreads);
+        currentLevel = std::make_shared<Level>(nThreads);
+    }
+
+    return currentLevel;
+}
+
+std::shared_ptr<Level> levelFromBatchRequest(
+  const std::shared_ptr<faabric::BatchExecuteRequest>& req)
+{
+    const auto other =
+      reinterpret_cast<const Level*>(req->contextdata().data());
+    auto lvl = std::make_shared<Level>(other->numThreads);
+    lvl->deserialise(other);
+
+    return lvl;
 }
 
 Level::Level(int numThreadsIn)
-  : numThreads(numThreadsIn)
-  , barrier(numThreadsIn)
+  : id(faabric::util::generateGid())
+  , numThreads(numThreadsIn)
 {}
 
+std::vector<uint32_t> Level::getSharedVars()
+{
+    return std::vector<uint32_t>(sharedVars, sharedVars + nSharedVars);
+}
+
+void Level::setSharedVars(uint32_t* ptr, int nVars)
+{
+    sharedVars = new uint32_t[nVars];
```

Comment:
First off `sharedVars` is probably not a great name as they're really wasm offsets, so `sharedVarOffsets` might be better. Because they're wasm offsets, they'll always be `uint32_t`.

I'm not sure what sort of abstraction would make things any safer or clearer here. It would have to be serialisable so would need to be composed of primitives itself, so it could be a struct that held the value and size, e.g. 

```
struct SharedVarPtr {
    uint32_t val;
    size_t size = sizeof(uint32_t);
}
```

but I'm not sure that adds much clarity or extra safety (plus the `size` would always be constant). I could add a `#define SHARED_VAR_PTR_SIZE  sizeof(uint32_t)` at the top to make that part explicit, but it's only used in two places so I'm not sure it's worth it.

Any mistakes in the serialisation and deserialisation should be picked up in the tests, so I'm happy to leave this as is.